### PR TITLE
ceph: add db-devices flag and fix DatabaseSizeMB

### LINKED ISSUE #3652: All non-rotational devices consumed as metadataDevice when only one was specified
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Ceph consumes all flash devices for block.db and wal when one specific device was configured as a `metadataDevice`.

**Expected behavior:**
Rook should configure Ceph to only use the device specified for block.db/wal.

**How to reproduce it (minimal and precise):**
Create a Rook Ceph cluster with two (or presumably more) non-rotational devices and one or more rotational devices, but only specify one of the flash devices to be used as the `metadataDevice` on the node.

**Other comments:**

```
c4tw8z1:~$ lsblk
NAME                                                MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda                                                   8:0    0 238.5G  0 disk
└─sda1                                                8:1    0 238.5G  0 part /etc/ssl/certs/ca-certificates.crt.rancher
sdb                                                   8:16   0 372.6G  0 disk
└─ceph--block--dbs--06206d70--22ce--4a20--b035--68ad7aea3317-osd--block--db--c84de295--a8dd--4b0d--b3eb--29841b462991
                                                    252:3    0   558G  0 lvm
sdc                                                   8:32   0   9.1T  0 disk
└─ceph--block--22045f1b--507a--4f88--b078--27170fa79f40-osd--block--8e3152a5--5669--411b--93a4--c747e66dc1c3
                                                    252:2    0   9.1T  0 lvm
sdd                                                   8:48   0   9.1T  0 disk
└─ceph--block--3f4df301--a0f8--451d--b343--bf8eeee4c0cb-osd--block--43362e5b--9784--47be--a2b9--1f644ded681d
                                                    252:0    0   9.1T  0 lvm
nvme0n1                                             259:0    0 745.2G  0 disk
├─ceph--block--dbs--06206d70--22ce--4a20--b035--68ad7aea3317-osd--block--db--6c3403e9--513c--4b6d--99ee--87a7a6653f44
│                                                   252:1    0   558G  0 lvm
└─ceph--block--dbs--06206d70--22ce--4a20--b035--68ad7aea3317-osd--block--db--c84de295--a8dd--4b0d--b3eb--29841b462991
                                                    252:3    0   558G  0 lvm
```

It occurs to me there may also be more that contributed to this behavior (at least in my case), or it may be an entirely separate issue, as it looks like Rook is also not merging the cluster-level `storage.config` with my node-level `storage.node.config`, so the `databaseSizeMB` setting is getting ignored, and the metadata lvs are getting created as 50% of the sum of both non-rotational devices since there are two rotational devices. Not sure if this is intended behavior.

Based on the comments, I believe this issue may be a regression introduced by #2777 due to the apparent omission of the `--db-devices` argument when the change was made to use batch provisioning with ceph-volume, but I can't be certain because I haven't tested with a version prior to this PR. In either case, there are lots of juicy comments on this PR specifically pertaining to this issue, which appears to have been known before the PR was accepted.

Other possibly related issues: #3092, #3449

**File(s)**:
* Cluster CRD: [cluster.zip](https://github.com/rook/rook/files/3520956/cluster.zip) (use c4tw8z1 as a reference node)

**Environment**:
* OS: Rancher OS 1.5.3
* Kernel: 4.14.128-rancher
* Hardware configuration: Dell R320, 800GB NVMe, 400GB SATA SSD, 2x 10TB HDD
* Rook version: 1.0.2
* Storage backend version: Ceph v14.2.2
* Kubernetes version: 1.13.5
* Kubernetes cluster type: bare metal w/ Rancher
* Storage backend status: healthy

### PR DESCRIPTION
**Pull request by mvollman**
Signed-off-by: Michael Vollman <michael.b.vollman@gmail.com>

**Description of your changes:**
Adding support for the db-devices flag for ceph versions 14.2.1 and
newer.  This flag will allow for explicitly setting the db device for
bluestore configurations to the device specified as metadataDevice.

Fixing the databaseSizeMB config parameter to ensure it is used to set
the size of the db-device partition when specified at either the global
or individual device level. (#3652)

**Which issue is resolved by this Pull Request:**
Resolves #3652 

**Checklist:**

- [X] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [X] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

### CODE REVIEW

File: pkg/daemon/ceph/osd/agent_test.go [hunk 1]
```diff
@@ -217,6 +217,11 @@ func testOSDAgentWithDevicesHelper(t *testing.T, storeConfig config.StoreConfig,
 			assert.Equal(t, config.Bluestore, storeConfig.StoreType)
 			return `SIZE="1234567890" TYPE="part"`, nil
 		}
+		if command == "ceph" {
```
Comment (leseb) [inline pkg/daemon/ceph/osd/agent_test.go]: why is this needed?

File: pkg/daemon/ceph/osd/volume.go [hunk 2]
```diff
@@ -151,6 +153,15 @@ func updateLVMConfig(context *clusterd.Context) error {
 }
 
 func (a *OsdAgent) initializeDevices(context *clusterd.Context, devices *DeviceOsdMapping) error {
+	versionString, versionErr := context.Executor.ExecuteCommandWithOutput(false, "", "ceph", "--version")
```
Comment (leseb) [inline pkg/daemon/ceph/osd/volume.go]: Can you use `err` instead of `versionErr`, it's more common.

File: pkg/daemon/ceph/osd/volume.go [hunk 3]
```diff
@@ -151,6 +153,15 @@ func updateLVMConfig(context *clusterd.Context) error {
...
+	if versionErr != nil {
+		return fmt.Errorf("failed to get ceph version. %+v", versionErr) // fail return here as validation provided by ceph-volume
+	}
+	version, extractErr := cephver.ExtractCephVersion(versionString)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/volume.go]: ditto

File: pkg/daemon/ceph/osd/volume.go (hunk 2 above)
Comment (leseb) [inline pkg/daemon/ceph/osd/volume.go]: Please don't do this, use `a` has cluster which has `ClusterInfo` and `CephVersion`. It should already have the correct Ceph version so let's use it.

**Comment by galexrt** [comment]
Restarting CI due to `no space left on device` errors.

File: pkg/daemon/ceph/osd/volume.go [hunk 4]
```diff
@@ -289,6 +306,14 @@ func (a *OsdAgent) initializeDevices(context *clusterd.Context, devices *DeviceO
 	return nil
 }
 
+func getDatabaseSize(globalSize int, deviceSize int) int {
+	dbSize := globalSize
```
Comment (leseb) [inline pkg/daemon/ceph/osd/volume.go]: Why not just using `globalSize`, why do you need a new variable? Also, we need test for this function.

**Comment by galexrt** [comment]
CI failed with:
```
2019-08-28 08:23:53.908609 I | cephcmd: desired devices to configure osds: [{Name:sda OSDsPerDevice:1 MetadataDevice: DatabaseSizeMB:0 IsFilter:false} {Name:sdb OSDsPerDevice:1 MetadataDevice: DatabaseSizeMB:0 IsFilter:false} {Name:nvme01 OSDsPerDevice:5 MetadataDevice: DatabaseSizeMB:0 IsFilter:false}]
2019-08-28 08:23:53.908726 I | cephcmd: desired devices to configure osds: [{Name:sdd OSDsPerDevice:1 MetadataDevice:sdb DatabaseSizeMB:2048 IsFilter:false} {Name:sde OSDsPerDevice:1 MetadataDevice:sdb DatabaseSizeMB:0 IsFilter:false} {Name:sdf OSDsPerDevice:1 MetadataDevice:sdc DatabaseSizeMB:0 IsFilter:false} {Name:sdg OSDsPerDevice:1 MetadataDevice:sdc DatabaseSizeMB:0 IsFilter:false}]
--- FAIL: TestParseDesiredDevices (0.00s)
    osd_test.go:63: 
        	Error Trace:	osd_test.go:63
        	Error:      	Not equal: 
        	            	expected: string("2048")
        	            	actual  : int(2048)
        	Test:       	TestParseDesiredDevices
    osd_test.go:64: 
        	Error Trace:	osd_test.go:64
        	Error:      	Not equal: 
        	            	expected: string("")
        	            	actual  : int(0)
        	Test:       	TestParseDesiredDevices
    osd_test.go:65: 
        	Error Trace:	osd_test.go:65
        	Error:      	Not equal: 
        	            	expected: string("")
        	            	actual  : int(0)
        	Test:       	TestParseDesiredDevices
    osd_test.go:66: 
        	Error Trace:	osd_test.go:66
        	Error:      	Not equal: 
        	            	expected: string("")
        	            	actual  : int(0)
        	Test:       	TestParseDesiredDevices
```

See https://jenkins.rook.io/blue/rest/organizations/jenkins/pipelines/rook/pipelines/rook/branches/PR-3696/runs/8/nodes/32/steps/33/log/?start=0

File: cmd/rook/ceph/osd.go [hunk 5]
```diff
@@ -275,9 +275,9 @@ func commonOSDInit(cmd *cobra.Command) {
 // Parse the devices, which are comma separated. A colon indicates a non-default number of osds per device
 // or a non collocated metadata device.
 // For example, one osd will be created on each of sda and sdb, with 5 osds on the nvme01 device.
-//   sda:1,sdb:1,nvme01:5
+//   sda:1::,sdb:1::,nvme01:5::
```
Comment (leseb) [inline cmd/rook/ceph/osd.go]: Can we add this in the documentation too?

File: pkg/daemon/ceph/osd/volume.go [hunk 6]
```diff
@@ -203,18 +192,26 @@ func (a *OsdAgent) initializeDevices(context *clusterd.Context, devices *DeviceO
...
-					metadataDevices[md] = []string{
-						osdsPerDeviceFlag,
-						deviceOSDCount,
-						deviceArg,
+					metadataDevices[md] = make(map[string]string)
+					metadataDevices[md]["osdsperdevice"] = deviceOSDCount
+					metadataDevices[md]["devices"] = deviceArg
+				}
+				deviceDBSizeMB := getDatabaseSize(a.storeConfig.DatabaseSizeMB, device.Config.DatabaseSizeMB)
+				if storeFlag == "--bluestore" && deviceDBSizeMB > 0 {
```
Comment (leseb) [inline pkg/daemon/ceph/osd/volume.go]: What if `deviceDBSizeMB` <= 0? Do we have validation for that somewhere?

File: cmd/rook/ceph/osd.go (hunk 5 above)
Comment (leseb) [reply cmd/rook/ceph/osd.go]: I only see this: https://rook.io/docs/rook/v1.0/ceph-cluster-crd.html#osd-configuration-settings which does not describe the `sda:1::,sdb:1::,nvme01:5::` behavior, or am I missing something?

**Comment by leseb** [comment]
I merged https://github.com/rook/rook/pull/3704 but this one is still a nice to have.

**Comment by mvollman** [comment]
Rebasing and adding PR #3593 to this PR so the two changes can be tested together.

File: cmd/rook/ceph/osd_test.go [hunk 7]
```diff
@@ -49,8 +49,8 @@ func TestParseDesiredDevices(t *testing.T) {
 	assert.Nil(t, result)
 	assert.NotNil(t, err)
 
-	// metadataDevice and OSDsPerDevice
-	devices = "sdd:1:sdb,sde:1:sdb,sdf:1:sdc,sdg:1:sdc"
+	// metadataDevice, databaseSizeMB, deviceClass and OSDsPerDevice
```
Comment (travisn) [inline cmd/rook/ceph/osd_test.go]: the order is this, right? osdsPerDevice, databaseSizeMB, deviceClass, metadataDevice

**Comment by travisn** [comment]
> Would it be possible to also include the support for several DB devices for several OSDs in the patch?

@mvollman Did you look at what it would take to support multiple db devices? Maybe that's a separate PR though.

**Comment by travisn** [comment]
@mvollman The unit test `TestParseDesiredDevices()` is failing: 
```
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 26 [running]:
testing.tRunner.func1(0xc000548f00)
	/usr/local/go/src/testing/testing.go:792 +0x387
panic(0x17634e0, 0x2b6d990)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/rook/rook/cmd/rook/ceph.TestParseDesiredDevices(0xc000548f00)
	/home/rook/go/src/github.com/rook/rook/cmd/rook/ceph/osd_test.go:55 +0x1192
```

File: cmd/rook/ceph/osd_test.go (hunk 7 above)
Comment (mvollman) [reply cmd/rook/ceph/osd_test.go]: Yes. I will fix

**Comment by mvollman** [comment]
I'm thinking separate PR for multiple md device support. Either the current parameter is extended to be a comma separated string or a new config option is introduced `metadataDeviceList` that is an array of md devices.  I will take a look at the unit test tonight.

**Comment by travisn** [comment]
Sounds great! I would vote for the comma separated string instead of a new property.

**Comment by christianhuening** [comment]
great stuff! Thank you! 
will this be backported to 1.0.6?

**Comment by leseb** [comment]
@christianhuening they are not bug fixes so they will be in 1.1. We do not backport features. Thanks.

**Comment by christianhuening** [comment]
this is not considered a bug? it even has the label "bug". For us this definitely breaks operations. So I would consider this a bug. Or did I get you wrong?

**Comment by travisn** [comment]
@christianhuening To be clear, which is the most critical part of this change for you? There were  really several changes in this PR. It is a bigger change so it's riskier to backport. I took a quick look at backporting and there are also a number of merge conflicts, which increases risk. If we could find a smaller fix to backport it would be more feasible.

**Comment by dimm0** [comment]
Removing the 20G default is the one for me... I want to add a node and have rook using the largest possible partitions

**Comment by christianhuening** [comment]
@travisn We updated from 0.9.3 to 1.0.5. Then we wanted to test our disk failover solution of removing an OSD and readding it. We got stuck doing this when readding, since rook now uses ceph-volume to format the devices and that collided with our 0.9.x formatted metadatadevice.  So we removed the complete node, wiped all the disks and attempted to rejoin the whole node. Then the rook prepare job fails at configuring the metadatadevice. The metadatadevie is not detected or taken from the configuration. So right now, we cannot re-add the node with our configuration.

**Comment by travisn** [comment]
@dimm0 Just opened #3737 for that one
@christianhuening i see, you really need the new --db-devices flag. @mvollman do you think there is a scoped fix we could backport to 1.0 to lower risk?

**Comment by christianhuening** [comment]
@travisn we can also try and upgrade to Nautilus. However, this is our development cluster. We're pretty concerned about how the update procedure will work out on the prod cluster. Not having to do the big leap all at once, would be much appreciated ;-)

**Comment by mvollman** [comment]
@travisn The commits were not structured in a way that allows to easily pick just the db-devices portion of this change.  Also the current code supports db-devices flag for 14.2.1 and newer.  If there are mimic versions that support db-devices the code would need to be updated to reflect that.  Right now db-devices will not be used with any mimic versions.

**Comment by travisn** [comment]
According to [this](https://github.com/rook/rook/issues/3652#issuecomment-526625007), we could include 13.2.6 in consideration for calling the --db-devices flag. 
@christianhuening v1.1 is around the corner. Could you wait for that release?

**Comment by christianhuening** [comment]
Around the corner sounds good. Next week?

**Comment by travisn** [comment]
The target date is Sept 10th, but always depends on if we find blocking issues during testing.
